python/lsst/daf/persistence/butler.py

```diff
@@ -1511,21 +1502,35 @@ def dataRef(self, datasetType, level=None, dataId={}, **rest):
         return ButlerDataRef(subset, subset.cache[0])
 
     def _read(self, location):
-        """Unpersist an object using data inside a butlerLocation object.
+        """Unpersist an object using data inside a ButlerLocation or ButlerComposite object.
 
         Parameters
         ----------
-        location - ButlerLocation
-            A butlerLocation instance populated with data needed to read the object.
+        location : ButlerLocation or ButlerComposite
+            A ButlerLocation or ButlerComposite instance populated with data needed to read the object.
 
         Returns
         -------
-        object - an instance of the object specified by the butlerLocation.
+        object
+            An instance of the object specified by the location.
         """
         self.log.debug("Starting read from %s", location)
-        results = location.repository.read(location)
-        if len(results) == 1:
-            results = results[0]
+
+        if isinstance(location, ButlerComposite):
+            for name, componentInfo in location.componentInfo.items():
+                if componentInfo.subset:
+                    subset = self.subset(datasetType=componentInfo.datasetType, dataId=location.dataId)
+                    componentInfo.obj = [obj.get() for obj in subset]
+                else:
+                    obj = self.get(componentInfo.datasetType, location.dataId, immediate=True)
+                    componentInfo.obj = obj
+                assembler = location.assembler or genericAssembler
+            results = assembler(dataId=location.dataId, componentInfo=location.componentInfo, cls=location.python)
+            return results
+        else:
+            results = location.repository.read(location)
+            if len(results) == 1:
+                results = results[0]
```

Comment:
The above assignment is changing the type for 'results' on the fly - consider using a new variable?